### DISCUSSION TITLE
Do we need chain id in every header?

### DISCUSSION THREAD

**Discussion by cmwaters** [tendermint/tendermint#5788] [Ideas]
A recent shower thought:  Do we need chain id in every header? 

Internally, every node has a record of the chain ID. They get it from genesis and it becomes an immutable part of state. Nodes interacting with one another through consensus or fast sync know that a header (and therefore a block) is part of their chain by using this internal chain ID to verify the signatures. In this way the chainID is already meshed in with the hashes that form the header and make it unique to the network.

Another way to look at it is if we had a twin network with the exact same validators and order of txs, if the genesis chain ID were different then we would still get different hashes for each of the headers even if the chainID itself wasn't in the header.

What I can imagine we need it for is an ergonomic feature for external clients that query the data over the RPC. But can't we then just append it in the messages we send.

**Comment by tac0turtle** [comment]
I like this idea! If we wanted to include the chainID in the header in some form we could hash it into one of the fields or add it to a tree of hashes when https://github.com/tendermint/spec/issues/199 is implemented

**Comment by alexanderbez** [comment]
Makes sense to me!
